include/lbann/weights/weights.hpp

```diff
@@ -43,57 +41,79 @@ namespace lbann {
 // Forward declaration
 class optimizer;
 
-/** Layer weights.
- *  Similar to Tensorflow "variables."
+/** Neural network weights.
+ *  Weights are tensors that act as trainable parameters for a neural
+ *  network. The values can be initialized with a weights initializer
+ *  and are optimized with first-order methods.
+ *
+ *  Internally, the weight values are stored in a distributed
+ *  matrix. The matrix height dimensions are tensor dimensions that
+ *  correspond to the matrix height. Similarly, the matrix width
+ *  dimensions correspond to the matrix width.
+ *
+ *  Note that LBANN weights are similar to Tensorflow variables and
+ *  Caffe parameters.
  */
 class weights {
-
   friend class optimizer;
 
  public:
-
-  /** Constructor. */
   weights(lbann_comm* comm,
           cudnn::cudnn_manager* cudnn = nullptr);
-
-  /** Copy constructor. */
   weights(const weights& other);
-  /** Copy assignment operator. */
   weights& operator=(const weights& other);
-  /** Destructor. */
   virtual ~weights();
 
   /** Set weights name.
    *  Each set of weights in a model should have a unique name.
    */
-  void set_name(std::string name) { m_name = name; }
-
+  inline void set_name(const std::string name) { m_name = name; }
   /** Get weights name. */
-  std::string get_name() const { return m_name; }
+  inline std::string get_name() const { return m_name; }
 
-  /** Create a copy of the weights. */
+  /** Create a copy of the weights.
+   *  This function dynamically allocates memory for a weights
+   *  instance and instantiates a copy. The caller is responsible for
+   *  deallocating the instance.
+   */
   virtual weights* copy() const { return new weights(*this); }
 
-  /** Setup weights. */
+  /** Setup weights as a vector. */
+  virtual void setup(int size);
+  /** Setup weights as a tensor. */
+  virtual void setup(std::vector<int> tensor_dims);
+  /** Setup weights as a matrix. */
   virtual void setup(int height,
                      int width,
                      El::Distribution col_dist,
                      El::Distribution row_dist);
-  /** Setup GPU objects for weights. */
-  virtual void setup_gpu();
+  /** Setup weights as a matrix with tensor height and width. */
+  virtual void setup(std::vector<int> matrix_height_dims,
+                     std::vector<int> matrix_width_dims,
+                     El::Distribution col_dist,
+                     El::Distribution row_dist);
 
-  /** Get height of weights matrix. */
-  int get_height() const { return m_height; }
-  /** Get width of weights matrix. */
-  int get_width() const { return m_width; }
+  /** Get weight tensor dimensions. */
+  std::vector<int> get_dims() const;
```

Comment:
Could you please provide a note/example as part of documentation as to what each entry of the vector return  get_dims() (weight tensor dimensions) may represent e.g., dims[0] is N, dims[1] is channel etc?

Reply:
I've expanded the documentation, although you should check if it is clear. Ultimately, this feature just implements a mathematical tensor and it is the responsibility of the layer class to interpret it.